lib/src/socket_connector.dart

```diff
@@ -421,27 +468,53 @@ class SocketConnector {
 
     int connections = 0;
     // bind to a local port for side 'A'
-    connector._serverSocketA = await ServerSocket.bind(addressA, portA);
-    // listen on the local port and connect the inbound socket
-    connector._serverSocketA?.listen((sideASocket) async {
-      if (!multi) {
-        unawaited(connector._serverSocketA?.close());
-      }
+    connector._serverSocketA = await ServerSocket.bind(
+      addressA,
+      portA,
+      backlog: backlog,
+    );
+
+    StreamController<Socket> ssc = StreamController();
+    ssc.stream.listen((sideASocket) async {
       Side sideA = Side(sideASocket, true, transformer: transformAtoB);
-      unawaited(connector.handleSingleConnection(sideA));
+      unawaited(connector.handleSingleConnection(sideA).catchError((err) {
+        logSink
+            .writeln('ERROR $err from handleSingleConnection on sideA $sideA');
+      }));
 
       if (verbose) {
-        logSink.writeln('Making connection ${++connections} to the "B" side');
+        logSink.writeln('Creating socket #${++connections} to the "B" side');
       }
       // connect to the side 'B' address and port
       Socket sideBSocket = await Socket.connect(addressB, portB);
+      if (verbose) {
+        logSink.writeln('"B" side socket #$connections created');
+      }
       Side sideB = Side(sideBSocket, false, transformer: transformBtoA);
-      beforeJoining?.call(sideA, sideB);
-      unawaited(connector.handleSingleConnection(sideB));
+      if (verbose) {
+        logSink.writeln('Calling the beforeJoining callback');
+      }
+      await beforeJoining?.call(sideA, sideB);
+      unawaited(connector.handleSingleConnection(sideB).catchError((err) {
+        logSink
+            .writeln('ERROR $err from handleSingleConnection on sideB $sideB');
+      }));
 
       onConnect?.call(sideASocket, sideBSocket);
     });
 
+    // listen on the local port and connect the inbound socket
+    connector._serverSocketA?.listen((sideASocket) {
```

Comment:
this closure now chucks the socket onto the other stream controller and returns immediately.